trie/branchNode.go

```diff
@@ -683,43 +625,44 @@ func (bn *branchNode) delete(
 		return false, bn
 	}
 
-	if len(oldHash) != 0 {
-		modifiedHashes.Append([][]byte{oldHash})
-	}
-	bn.dirty = true
+	return bn.reduceNodeIfNeeded(goRoutinesManager, modifiedHashes, db)
+}
+
+func (bn *branchNode) reduceNodeIfNeeded(
+	goRoutinesManager common.TrieGoroutinesManager,
+	modifiedHashes common.AtomicBytesSlice,
+	db common.TrieStorageInteractor,
+) (bool, node) {
+	bn.mutex.Lock()
+	defer bn.mutex.Unlock()
```

Comment:
if setters in baseNode are protected by mutex, i think this protection here is not needed

Reply:
It is needed in order to keep all changes atomic. If I remove this there might be edgecases where different baseNode fields change from different goroutines.